### PR TITLE
feat: Animations in LED Badge Preview

### LINKED ISSUE #93: Animations (Marquee, Flash) in Badge Preview of the App
**Description**
The Badge should react according to the selected animation. Currently it only displays the Text onto the preview.

**GIF of How it should work**
temp

**Would you like to work on the issue?**
YES

### PR DESCRIPTION
**Pull request by adityastic**
Implemented the Four Basic Animations

Fixes #93 

Changes: 
- Added Animations for Modes, with Speed Control, which also includes Features of Marquee and Flash
- ValueAnimator Manages the Values for the animation
- Control Is Based on the Type of the animation mode from now

Screenshots for the change:
ezgif com-video-to-gif
ezgif com-video-to-gif-2

Special Thanks to @yashk2000 for providing me the exact hardware animations

### CODE REVIEW

**Comment by adityastic** [comment]
@iamareebjamal Please have a look at this

**Comment by iamareebjamal** [comment]
Ask for peer review first

**Comment by adityastic** [comment]
@iamareebjamal sure sir. @yashk2000 can you please review it

**Comment by yatri1609** [comment]
@adityastic why is there the outline on the screen in case of marquee?

**Comment by adityastic** [comment]
Because it is what marquee represents on the actual LED Badge. Here is a Video of the Hardware Programmed with Marquee and Mode (Fixed) :- https://drive.google.com/file/d/1SL1w_jlEGGNbDfUUSo69BNPxi1_LjEqX/view?usp=drivesdk

**Comment by yatri1609** [comment]
@adityastic ohk but marquee as per name suggests text moving in a particular direction. Anyways it's like that here than it's fine !

**Comment by adityastic** [comment]
@yatri1609 Yes, Even I got confused. @yashk2000 gave me the actual LED footage as I do not own the badge myself and I just formulated it

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 1]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
         // Draw Cells
         for (i in 0 until badgeHeight) {
             for (j in 0 until badgeWidth) {
-                if (i < checkList?.size ?: -1 && j < checkList?.get(i)?.list?.size ?: -1 && checkList?.get(i)?.list?.get(j) == true) {
-                    ledEnabled?.bounds = cells[i].list[j]
-                    ledEnabled?.draw(canvas)
+                if (ifFlash == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Please remove `if (ifFlash == true)`, `if (ifFlash)` is enough.

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 2]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+                    val aI: Int = animationIndex % 800
+                    val valid: Boolean = aI > 400
+
+                    if (valid && i < checkList?.size ?: -1 && j < checkList?.get(i)?.list?.size ?: -1 && checkList?.get(i)?.list?.get(j) == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove the `== true` from here too.

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 3]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+                        ledEnabled?.bounds = cells[i].list[j]
+                        ledEnabled?.draw(canvas)
+                    } else {
+                        ledDisabled?.bounds = cells[i].list[j]
+                        ledDisabled?.draw(canvas)
+                    }
+                } else if (ifMarquee == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove `== true`

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 4]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+                            (i + j - 1) % 4 == (3 - (aI % 4))
+                        }
+                    } else {
+                        false
+                    }
+
+                    when (badgeMode) {
+                        Mode.LEFT -> {
+                            val animationValue = animationIndex.div(200)
+                            if (valid || i < checkList?.size ?: -1 && j < checkList?.get(i)?.list?.size ?: -1 && j >= animationValue && checkList?.get(i)?.list?.get(j - animationValue) == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove `== true`

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 5]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+                                ledEnabled?.bounds = cells[i].list[j]
+                                ledEnabled?.draw(canvas)
+                            } else {
+                                ledDisabled?.bounds = cells[i].list[j]
+                                ledDisabled?.draw(canvas)
+                            }
+                        }
+                        Mode.RIGHT -> {
+                            val animationValue = animationIndex.div(200)
+                            if (valid || i < checkList?.size ?: -1 && j < checkList?.get(i)?.list?.size ?: -1 && j <= (43 - animationValue) && checkList?.get(i)?.list?.get((animationValue.plus(j))) == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove `== true`

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 6]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+                                ledEnabled?.bounds = cells[i].list[j]
+                                ledEnabled?.draw(canvas)
+                            } else {
+                                ledDisabled?.bounds = cells[i].list[j]
+                                ledDisabled?.draw(canvas)
+                            }
+                        }
+                        Mode.UP -> {
+                            val animationValue = animationIndex.div(800)
+                            if (valid || i < checkList?.size ?: -1 && j < checkList?.get(i)?.list?.size ?: -1 && i >= animationValue && checkList?.get(i - animationValue)?.list?.get(j) == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove `== true`

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 7]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+                                ledEnabled?.bounds = cells[i].list[j]
+                                ledEnabled?.draw(canvas)
+                            } else {
+                                ledDisabled?.bounds = cells[i].list[j]
+                                ledDisabled?.draw(canvas)
+                            }
+                        }
+                        Mode.DOWN -> {
+                            val animationValue = animationIndex.div(800)
+                            if (valid || i < checkList?.size ?: -1 && j < checkList?.get(i)?.list?.size ?: -1 && i <= (10 - animationValue) && checkList?.get(animationValue.plus(i))?.list?.get(j) == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove `== true`

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 8]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+                                ledEnabled?.bounds = cells[i].list[j]
+                                ledEnabled?.draw(canvas)
+                            } else {
+                                ledDisabled?.bounds = cells[i].list[j]
+                                ledDisabled?.draw(canvas)
+                            }
+                        }
+                        Mode.FIXED -> {
+                            if ((valid) || i < checkList?.size ?: -1 && j < checkList?.get(i)?.list?.size ?: -1 && checkList?.get(i)?.list?.get(j) == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove `== true`

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 9]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+                        Mode.LASER -> {
+                        }
+                    }
                 } else {
-                    ledDisabled?.bounds = cells[i].list[j]
-                    ledDisabled?.draw(canvas)
+                    when (badgeMode) {
+                        Mode.LEFT -> {
+                            val animationValue = animationIndex.div(200)
+                            if (i < checkList?.size ?: -1 && j < checkList?.get(i)?.list?.size ?: -1 && j >= animationValue && checkList?.get(i)?.list?.get(j - animationValue) == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove `== true`

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 10]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+                                ledEnabled?.bounds = cells[i].list[j]
+                                ledEnabled?.draw(canvas)
+                            } else {
+                                ledDisabled?.bounds = cells[i].list[j]
+                                ledDisabled?.draw(canvas)
+                            }
+                        }
+                        Mode.RIGHT -> {
+                            val animationValue = animationIndex.div(200)
+                            if (i < checkList?.size ?: -1 && j < checkList?.get(i)?.list?.size ?: -1 && j <= (43 - animationValue) && checkList?.get(i)?.list?.get((animationValue.plus(j))) == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove `== true`

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 11]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+                                ledEnabled?.bounds = cells[i].list[j]
+                                ledEnabled?.draw(canvas)
+                            } else {
+                                ledDisabled?.bounds = cells[i].list[j]
+                                ledDisabled?.draw(canvas)
+                            }
+                        }
+                        Mode.UP -> {
+                            val animationValue = animationIndex.div(800)
+                            if (i < checkList?.size ?: -1 && j < checkList?.get(i)?.list?.size ?: -1 && i >= animationValue && checkList?.get(i - animationValue)?.list?.get(j) == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove `== true`

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 12]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+                                ledEnabled?.bounds = cells[i].list[j]
+                                ledEnabled?.draw(canvas)
+                            } else {
+                                ledDisabled?.bounds = cells[i].list[j]
+                                ledDisabled?.draw(canvas)
+                            }
+                        }
+                        Mode.DOWN -> {
+                            val animationValue = animationIndex.div(800)
+                            if (i < checkList?.size ?: -1 && j < checkList?.get(i)?.list?.size ?: -1 && i <= (10 - animationValue) && checkList?.get(animationValue.plus(i))?.list?.get(j) == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove `== true`

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt (hunk 1 above)
Comment (adityastic) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: ifFlash is a nullable property, it needs to be checked against its value.

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 13]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+                                ledEnabled?.bounds = cells[i].list[j]
+                                ledEnabled?.draw(canvas)
+                            } else {
+                                ledDisabled?.bounds = cells[i].list[j]
+                                ledDisabled?.draw(canvas)
+                            }
+                        }
+                        Mode.FIXED -> {
+                            if (i < checkList?.size ?: -1 && j < checkList?.get(i)?.list?.size ?: -1 && checkList?.get(i)?.list?.get(j) == true) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove `== true`

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt (hunk 2 above)
Comment (adityastic) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Same, Sir. Nullable Property, won't work without checking against the value

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 14]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+            Speed.EIGHT -> 8
+        }
+
+        configValueAnimation()
 
-        if (allHex.size > 0 && 8 * allHex.size < badgeWidth) {
-            val diff = ((badgeWidth - (8 * allHex.size)) / 2)
+        val diff: Int
+        if (allHex.size > 0) {
+            diff = ((badgeWidth - (8 * allHex.size)) / 2)
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove the magic numbers.

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 15]
```diff
@@ -95,30 +108,216 @@ class PreviewBadge : View {
...
+            if (8 * allHex.size < badgeWidth) {
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Remove the magic number 8.

File: app/src/main/java/com/nilhcem/blenamebadge/ui/message/MessageActivity.kt [hunk 16]
```diff
@@ -61,7 +61,7 @@ class MessageActivity : AppCompatActivity() {
         mode.adapter = ArrayAdapter<String>(this, spinnerItem, Mode.values().map { getString(it.stringResId) })
 
         send.setOnClickListener {
-            val inputManager: InputMethodManager = this?.getSystemService(Context.INPUT_METHOD_SERVICE)
+            val inputManager: InputMethodManager = this.getSystemService(Context.INPUT_METHOD_SERVICE)
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/message/MessageActivity.kt]: Was this really needed?

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt (hunk 2 above)
Comment (yashk2000) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: If that's the case, then ok. Resolve similar comments then.

File: app/src/main/java/com/nilhcem/blenamebadge/ui/message/MessageActivity.kt (hunk 16 above)
Comment (adityastic) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/message/MessageActivity.kt]: Yes, This was called with a safe call keyword. But no safe call was needed as this variable is not nullable

**Comment by yashk2000** [comment]
Looks cool :+1:

**Comment by adityastic** [comment]
@iamareebjamal Peer Review Complete, Have a look

**Comment by adityastic** [comment]
@iamareebjamal Peer Review is done

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 17]
```diff
@@ -26,7 +30,17 @@ class PreviewBadge : View {
     private var badgeHeight = 11
     private var badgeWidth = 44
 
+    private var oneByte = 8
+
+    private var ifFlash: Boolean? = false
```
Comment (iamareebjamal) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Why nullable?

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 18]
```diff
@@ -26,7 +30,17 @@ class PreviewBadge : View {
...
+    private var ifMarquee: Boolean? = false
```
Comment (iamareebjamal) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Same?

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 19]
```diff
@@ -26,7 +30,17 @@ class PreviewBadge : View {
...
+    private var badgeSpeed: Int = 1
+    private var badgeMode: Mode = Mode.LEFT
+
+    private var animationIndex: Int = 0
+
     private var checkList: ArrayList<CheckList>? = null
```
Comment (iamareebjamal) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Why nullable?

**Comment by adityastic** [comment]
@iamareebjamal Done, Have a look Sir

**Comment by adityastic** [comment]
@iamareebjamal please have a look if any other issues are present

File: app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt [hunk 20]
```diff
@@ -15,48 +15,59 @@ import com.nilhcem.blenamebadge.R
...
+    private var oneByte = 8
+
+    private var ifFlash: Boolean = false
+    private var ifMarquee: Boolean = false
+    private var badgeSpeed: Int = 1
+    private var badgeMode: Mode = Mode.LEFT
+
+    private var animationIndex: Int = 0
+
+    private lateinit var checkList: ArrayList<CheckList>
```
Comment (iamareebjamal) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Same question, why lateinit at all? Why can't it be initialized here?

Comment (adityastic) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/badge_preview/PreviewBadge.kt]: Alright Sir

**Comment by adityastic** [comment]
@iamareebjamal any other changes, sir?

**Pull request merged**